File: sdk/csharp/ShellStorage.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Text;
using System.Collections.Generic; // Added missing import for Dictionary

namespace TuskLang
{
    /// <summary>
    /// Shell format storage for binary data
    /// Handles compression and binary serialization of TSK data
    /// </summary>
    public class ShellStorage
    {
        /// <summary>
        /// Magic bytes for Shell format identification
        /// </summary>
        public static readonly byte[] MAGIC = Encoding.UTF8.GetBytes("FLEX");
        
        /// <summary>
        /// Current version of Shell format
        /// </summary>
        public static readonly byte VERSION = 1;

        /// <summary>
        /// Pack data into Shell binary format
        /// </summary>
        public static byte[] Pack(ShellData data)
        {
            // Compress data if it's not already compressed
            byte[] compressedData;
            if (data.Data is string strData)
            {
                compressedData = CompressString(strData);
            }
            else if (data.Data is byte[] byteData)
            {
                compressedData = CompressBytes(byteData);
            }
            else
            {
                compressedData = CompressString(data.Data.ToString());
            }

            var idBytes = Encoding.UTF8.GetBytes(data.Id);

            // Build binary format
            // Magic (4) + Version (1) + ID Length (4) + ID + Data Length (4) + Data
            using (var ms = new MemoryStream())
            using (var writer = new BinaryWriter(ms))
            {
                writer.Write(MAGIC);
                writer.Write(VERSION);
                writer.Write(idBytes.Length);
                writer.Write(idBytes);
                writer.Write(compressedData.Length);
                writer.Write(compressedData);

                return ms.ToArray();
            }
        }

        /// <summary>
        /// Unpack Shell binary format
        /// </summary>
        public static ShellData Unpack(byte[] shellData)
        {
            using (var ms = new MemoryStream(shellData))
            using (var reader = new BinaryReader(ms))
            {
                // Check magic bytes
                var magic = reader.ReadBytes(4);
                if (!magic.SequenceEqual(MAGIC))
                    throw new ArgumentException("Invalid shell format");

                // Read version
                var version = reader.ReadByte();
                if (version != VERSION)
                    throw new ArgumentException($"Unsupported shell version: {version}");

                // Read ID
                var idLength = reader.ReadInt32();
                var storageId = Encoding.UTF8.GetString(reader.ReadBytes(idLength));

                // Read data
                var dataLength = reader.ReadInt32();
                var compressedData = reader.ReadBytes(dataLength);

                // Decompress
                var data = DecompressBytes(compressedData);

                return new ShellData
                {
                    Version = version,
                    Id = storageId,
                    Data = data
                };
            }
        }

        /// <summary>
        /// Compress string data
        /// </summary>
        private static byte[] CompressString(string data)
        {
            var bytes = Encoding.UTF8.GetBytes(data);
            return CompressBytes(bytes);
        }

        /// <summary>
        /// Compress byte array data
        /// </summary>
        private static byte[] CompressBytes(byte[] data)
        {
            using (var ms = new MemoryStream())
            {
                using (var gzip = new GZipStream(ms, CompressionMode.Compress))
                {
                    gzip.Write(data, 0, data.Length);
                }
                return ms.ToArray();
            }
        }

        /// <summary>
        /// Decompress byte array data
        /// </summary>
        private static string DecompressBytes(byte[] compressedData)
        {
            using (var ms = new MemoryStream(compressedData))
            using (var gzip = new GZipStream(ms, CompressionMode.Decompress))
            using (var resultMs = new MemoryStream())
            {
                gzip.CopyTo(resultMs);
                return Encoding.UTF8.GetString(resultMs.ToArray());
            }
        }

        /// <summary>
        /// Detect the type of data stored in Shell format
        /// </summary>
        public static string DetectType(byte[] shellData)
        {
            try
            {
                var data = Unpack(shellData);
                var content = data.Data.ToString();

                // Try to detect JSON
                if (content.StartsWith("{") || content.StartsWith("["))
                    return "json";

                // Try to detect TSK format
                if (content.Contains("[") && content.Contains("]") && content.Contains("="))
                    return "tsk";

                // Try to detect XML
                if (content.StartsWith("<") && content.Contains(">"))
                    return "xml";

                // Default to text
                return "text";
            }
            catch
            {
                return "unknown";
            }
        }

        /// <summary>
        /// Create Shell data with metadata
        /// </summary>
        public static ShellData CreateShellData(object data, string id, Dictionary<string, object> metadata = null)
        {
            return new ShellData
            {
                Version = VERSION,
                Id = id,
                Data = data,
                Metadata = metadata ?? new Dictionary<string, object>()
            };
        }
    }

    /// <summary>
    /// Data structure for Shell storage
    /// </summary>
    public class ShellData
    {
        public byte Version { get; set; }
        public string Id { get; set; }
        public object Data { get; set; }
        public Dictionary<string, object> Metadata { get; set; } = new Dictionary<string, object>();
    }
} 

File: sdk/ruby/shell_storage.rb
# frozen_string_literal: true

# Shell Storage for Ruby
# Handles binary data compression and storage

module TuskLang
  # Shell format storage for binary data
  class ShellStorage
    # Magic bytes for Shell format identification
    MAGIC = 'FLEX'.bytes
    VERSION = 1

    # Pack data into Shell binary format
    def self.pack(data)
      # Compress data if it's not already compressed
      compressed_data = if data[:data].is_a?(String)
                          Zlib::Deflate.deflate(data[:data])
                        elsif data[:data].is_a?(Array)
                          Zlib::Deflate.deflate(data[:data].pack('C*'))
                        else
                          Zlib::Deflate.deflate(data[:data].to_s)
                        end

      id_bytes = data[:id].bytes

      # Build binary format
      # Magic (4) + Version (1) + ID Length (4) + ID + Data Length (4) + Data
      result = []
      result.concat(MAGIC)
      result << VERSION
      result.concat([id_bytes.length].pack('>N').bytes)
      result.concat(id_bytes)
      result.concat([compressed_data.length].pack('>N').bytes)
      result.concat(compressed_data.bytes)

      result.pack('C*')
    end

    # Unpack Shell binary format
    def self.unpack(shell_data)
      offset = 0

      # Check magic bytes
      magic = shell_data.bytes[offset, 4]
      raise ArgumentError, 'Invalid shell format' unless magic == MAGIC
      offset += 4

      # Read version
      version = shell_data.bytes[offset]
      raise ArgumentError, "Unsupported shell version: #{version}" unless version == VERSION
      offset += 1

      # Read ID
      id_length = shell_data.bytes[offset, 4].pack('C*').unpack('>N')[0]
      offset += 4
      storage_id = shell_data.bytes[offset, id_length].pack('C*').force_encoding('UTF-8')
      offset += id_length

      # Read data
      data_length = shell_data.bytes[offset, 4].pack('C*').unpack('>N')[0]
      offset += 4
      compressed_data = shell_data.bytes[offset, data_length].pack('C*')

      # Decompress
      data = Zlib::Inflate.inflate(compressed_data).force_encoding('UTF-8')

      {
        version: version,
        id: storage_id,
        data: data
      }
    end

    # Detect the type of data stored in Shell format
    def self.detect_type(shell_data)
      data = unpack(shell_data)
      content = data[:data].to_s

      # Try to detect JSON
      return 'json' if content.start_with?('{') || content.start_with?('[')

      # Try to detect TSK format
      return 'tsk' if content.include?('[') && content.include?(']') && content.include?('=')

      # Try to detect XML
      return 'xml' if content.start_with?('<') && content.include?('>')

      # Default to text
      'text'
    rescue
      'unknown'
    end

    # Create Shell data with metadata
    def self.create_shell_data(data, id, metadata = nil)
      {
        version: VERSION,
        id: id,
        data: data,
        metadata: metadata || {}
      }
    end
  end
end 